src/config/config-factory.ts

```diff
@@ -1,24 +1,107 @@
-import { PartnerConfig } from '.';
+import { ConfigSystem, PartnerConfig } from '.';
 import kenyaConfig from './chis-ke';
 import togoConfig from './chis-tg';
 import civConfig from './chis-civ';
 import maliConfig from './chis-ml';
+import path from 'path';
+import fs from 'fs';
 
-export const CONFIG_MAP: { [key: string]: PartnerConfig } = {
-  'CHIS-KE': kenyaConfig,
-  'CHIS-TG': togoConfig,
-  'CHIS-CIV': civConfig,
-  'CHIS-ML': maliConfig
-};
-
-export default function getConfigByKey(key: string = 'CHIS-KE'): PartnerConfig {
-  const usingKey = key.toUpperCase();
-  console.log(`Using configuration: ${key}`);
-  const result = CONFIG_MAP[usingKey];
-  if (!result) {
-    const available = JSON.stringify(Object.keys(CONFIG_MAP));
-    throw Error(`Failed to start: Cannot find configuration "${usingKey}". Configurations available are ${available}`);
+export default class ConfigFactory<T = Record<string, any>> {
+  private config: PartnerConfig | null = null;
+  private static configFactoryInstance: ConfigFactory;
+  private filePath: string;
+
+  private DEFAULT_CONFIG_MAP: { [key: string]: PartnerConfig } = {
+    'CHIS-KE': kenyaConfig,
+    'CHIS-TG': togoConfig,
+    'CHIS-CIV': civConfig,
+    'CHIS-ML': maliConfig
+  };
+
+  constructor() {
+    this.filePath = path.join(this.getConfigUploadDirectory(), 'config.json');
+  }
+
+  static getConfigFactory<T = Record<string, any>>(): ConfigFactory<T> {
+    if (!ConfigFactory.configFactoryInstance) {
+      ConfigFactory.configFactoryInstance = new ConfigFactory();
+    }
+
+    return ConfigFactory.configFactoryInstance as ConfigFactory<T>;
   }
 
-  return result;
+  loadConfig(key?: string): PartnerConfig {
+
+    if (this.config) {
+      return this.config;
+    }
+
+    if (fs.existsSync(this.filePath)) {
+      this.config = this.readConfig();
+      console.log(`Using uploaded configuration: ${this.filePath}`);
+      return this.config;
+    }
+
+    //set default config
+    if (!key) {
+      console.log('no configuration found. Setting to default');
+      const defaultConfig = kenyaConfig;
+      defaultConfig.config.domains = [{
```

Comment:
Wouldn't this override the domains in this config? Would it be better to just crash the app if the key isn't set in the environment variables since that would be a deployment config issue?

Reply:
The crash scenario is just below this code block which i think is a better way of handling this

Reply:
i have added default config in case env variable is not provided this can then be overridden via api

Reply:
environment variables should always be provided and they are checked at start up. I cannot recommend changing env variables via API. We also shouldn't be changing the config values in code

Reply:
thats not what im doing 
i simply making the env variable optional
if not provided the app falls back to a default config which allows overwriting via api with any auth domain
other configs ie env and uploaded will authenticate domain before overwriting 
chck it out